File: 2022/Rust/src/day02.rs
pub struct Game {
    opponent: Play,
    strategy: Strategy,
}

#[derive(Clone, Copy)]
pub enum Play {
    Rock = 1,
    Paper = 2,
    Scissors = 3,
}

#[derive(Clone, Copy)]
pub enum Strategy {
    X,
    Y,
    Z,
}

impl Play {
    fn score_against(self, other: Self) -> usize {
        match (other, self) {
            (Play::Scissors, Play::Rock)
            | (Play::Rock, Play::Paper)
            | (Play::Paper, Play::Scissors) => 6,
            (Play::Scissors, Play::Scissors)
            | (Play::Rock, Play::Rock)
            | (Play::Paper, Play::Paper) => 3,
            _ => 0,
        }
    }

    const fn loses_to(self) -> Self {
        match self {
            Play::Rock => Play::Paper,
            Play::Paper => Play::Scissors,
            Play::Scissors => Play::Rock,
        }
    }

    const fn draws(self) -> Self {
        self
    }

    const fn beats(self) -> Self {
        self.loses_to().loses_to()
    }
}

pub fn parse(input: &str) -> Vec<Game> {
    input
        .lines()
        .map(|line| {
            let mut chars = line.chars();
            let opponent = chars.next().unwrap();
            let _ = chars.next().unwrap();
            let strategy = chars.next().unwrap();

            let opponent = match opponent {
                'A' => Play::Rock,
                'B' => Play::Paper,
                'C' => Play::Scissors,
                _ => unreachable!(),
            };
            let strategy = match strategy {
                'X' => Strategy::X,
                'Y' => Strategy::Y,
                'Z' => Strategy::Z,
                _ => unreachable!(),
            };
            Game { opponent, strategy }
        })
        .collect()
}

pub fn part1(parsed: &[Game]) -> usize {
    parsed
        .iter()
        .map(|&Game { opponent, strategy }| {
            let to_play = match strategy {
                Strategy::X => Play::Rock,
                Strategy::Y => Play::Paper,
                Strategy::Z => Play::Scissors,
            };

            to_play as usize + to_play.score_against(opponent)
        })
        .sum()
}

pub fn part2(parsed: &[Game]) -> usize {
    parsed
        .iter()
        .map(|&Game { opponent, strategy }| {
            let to_play = match strategy {
                Strategy::X => opponent.beats(),
                Strategy::Y => opponent.draws(),
                Strategy::Z => opponent.loses_to(),
            };

            (to_play as usize) + to_play.score_against(opponent)
        })
        .sum()
}

pub fn run(input: &str) {
    let parsed = parse(input);
    println!("Part 1: {}", part1(&parsed));
    println!("Part 2: {}", part2(&parsed));
}

#[cfg(test)]
mod tests {
    use super::*;

    const INPUT: &str = "A Y
B X
C Z";

    #[test]
    fn test_part1() {
        let parsed = parse(INPUT);
        assert_eq!(15, part1(&parsed));
    }

    #[test]
    fn test_part2() {
        let parsed = parse(INPUT);
        assert_eq!(12, part2(&parsed));
    }
}

File: 2022/Haskell/src/Day02.hs
module Day02 (day02) where

import AoC
import Data.Text (Text)
import qualified Data.Text as Text

data Play = Rock | Paper | Scissors deriving (Show, Eq)

data Suggestion = X | Y | Z deriving (Show, Eq)

data Game = Game
  { opponent :: Play,
    suggestion :: Suggestion
  }
  deriving (Show, Eq)

data Outcome = Win | Lose | Draw deriving (Show, Eq)

type Parsed = [Game]

pPlay :: Char -> Maybe Play
pPlay x = case x of
  'A' -> Just Rock
  'B' -> Just Paper
  'C' -> Just Scissors
  _ -> Nothing

pSuggestion :: Char -> Maybe Suggestion
pSuggestion x = case x of
  'X' -> Just X
  'Y' -> Just Y
  'Z' -> Just Z
  _ -> Nothing

pPair :: [Char] -> Maybe Game
pPair [a, b] = case (pPlay a, pSuggestion b) of
  (Just opponent, Just suggestion) -> return Game {..}
  _ -> Nothing
pPair _ = Nothing

matchScore :: Play -> Play -> Int
matchScore a b = case outcomeOf a b of
  Win -> 6
  Draw -> 3
  Lose -> 0
  where
    beats :: Play -> Play -> Bool
    a `beats` b = case (a, b) of
      (Rock, Scissors) -> True
      (Paper, Rock) -> True
      (Scissors, Paper) -> True
      _ -> False

    outcomeOf :: Play -> Play -> Outcome
    outcomeOf a b
      | a == b = Draw
      | a `beats` b = Win
      | otherwise = Lose

playScore :: Play -> Int
playScore a = case a of
  Rock -> 1
  Paper -> 2
  Scissors -> 3

part1' :: Monad m => Game -> m Int
part1' Game {..} = do
  let toPlay = case suggestion of
        X -> Rock
        Y -> Paper
        Z -> Scissors
  return $ playScore toPlay + matchScore toPlay opponent

part2' :: Monad m => Game -> m Int
part2' Game {..} = do
  let toPlay = case suggestion of
        X -> willLose opponent
        Y -> willDraw opponent
        Z -> willBeat opponent
  return $ playScore toPlay + matchScore toPlay opponent
  where
    willBeat :: Play -> Play
    willBeat a = case a of
      Rock -> Paper
      Paper -> Scissors
      Scissors -> Rock

    willDraw :: Play -> Play
    willDraw = id

    willLose :: Play -> Play
    willLose a = case a of
      Rock -> Scissors
      Paper -> Rock
      Scissors -> Paper

day02 :: Runner Parsed Int
day02 =
  let year = 2022
      day = 2
      parser :: Text -> Maybe Parsed
      parser input = traverse pPair $ fmap head . words <$> lines (Text.unpack input)
      part1 :: Parsed -> Maybe Int
      part1 = fmap sum . traverse part1'
      part2 = fmap sum . traverse part2'
   in Runner {..}